### PR TITLE
feat/AB#104315_EIS-application-path-implementation

### PR DESCRIPTION
**Pull request by unai-reliefapp**
# Description

feat: integrate application shortcut as valid path for frontend 
feat: include resolver for application path in order to fetch correct data for activated route 
feat: add method to extract correct path for give application for the aforementioned reason 
feat: include shortcut value for user query in order to get correct path from given application 
feat: remove useless properties

## Useful links

- [Link to ticket](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/104315)
- [Link to back-end branch](https://github.com/ReliefApplications/ems-backend/pull/1123)
- Please insert any useful link ( documentation you used for example )

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Test A
- [ ] Test B

## Screenshots

Back office configuration
https://github.com/user-attachments/assets/1098b359-dc74-471f-9108-eec06680c869

Front office behavior
https://github.com/user-attachments/assets/8d2670ac-da9b-426f-941b-a4f553dbf56a

# Checklist:

( * == Mandatory ) 

- [X] * I have set myself as assignee of the pull request
- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have put the ticket for review, adding the oort-frontend team to the list of reviewers
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# More explanation
https://www.loom.com/share/05a716d61b9744faaf51fb304c21d1e5?sid=f87cf896-582a-4f76-93ae-8ceed801b145

### CODE REVIEW

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

please continue to work on your current ticket, I'm putting info there FYI and to keep notes when I'll go back to it

A few things about this PR:
- it seems the navigation using id is not working anymore in front-end, when a shortcut is configured
- navigation using shortcut is not working in back-office ( not sure it was part of the scope )
- if the request to get the application fails, user only sees a loading screen ( both back-office & front-office )
- when access is not provided / application cannot be found, we can use a similar approach that what we did for the web-widgets integration ( see 'app-error' component in web-widgets application ). We should only redirect the user to a default application if there is no id / shortcut in the url. If there is one but the loading of the app fails, we display the error message
- also, please check that elements ( editor / summary card / grid widgets ) that can build links to other pages, would work with the new url ( I think it's using the id, even if a shortcut is used )

- lastly, I'm putting this one in a different list as it's certainly the most important, you may have done some extra work by doing a specific query to get the shortcut in the router. Instead, you could edit the back-end application.query, to have a new shortcut argument, and find the application based on the shortcut & id ( so the query, in the front-end, would pass id & shortcut, both with the same value, coming from the url, and then, the back-end will find the according application ). It could save a lot of code

**Comment by unai-reliefapp** [comment]
> @unai-reliefapp
> 
> please continue to work on your current ticket, I'm putting info there FYI and to keep notes when I'll go back to it
> 
> A few things about this PR:
> 
> * it seems the navigation using id is not working anymore in front-end, when a shortcut is configured
> * navigation using shortcut is not working in back-office ( not sure it was part of the scope )
> * if the request to get the application fails, user only sees a loading screen ( both back-office & front-office )
> * when access is not provided / application cannot be found, we can use a similar approach that what we did for the web-widgets integration ( see 'app-error' component in web-widgets application ). We should only redirect the user to a default application if there is no id / shortcut in the url. If there is one but the loading of the app fails, we display the error message
> * also, please check that elements ( editor / summary card / grid widgets ) that can build links to other pages, would work with the new url ( I think it's using the id, even if a shortcut is used )
> * lastly, I'm putting this one in a different list as it's certainly the most important, you may have done some extra work by doing a specific query to get the shortcut in the router. Instead, you could edit the back-end application.query, to have a new shortcut argument, and find the application based on the shortcut & id ( so the query, in the front-end, would pass id & shortcut, both with the same value, coming from the url, and then, the back-end will find the according application ). It could save a lot of code

About the last point:

You mean that every time we fetch an application, check that the value is the shortcut and use the related id, or use the default value in the id parameter for the query if it does not find it?

File: apps/front-office/src/app/app-routing.module.ts [hunk 1]
```diff
@@ -1,7 +1,55 @@
...
+/**
+ * Resolve application id
+ * Application can use shortcut values to generate the url
+ * We check the given params for the application route and check if it's a shortcut
+ * If it's a shortcut, return the related application id, if not return current param
+ *
+ * @param route ActivatedRouteSnapshot
+ * @returns resolved application id
+ */
+const applicationIdResolver: ResolveFn<any> = async (
```
Comment (AntoineRelief) [inline apps/front-office/src/app/app-routing.module.ts]: @unai-reliefapp 

about my last comment, I believe this part is too complex, and you could easily solve it by just sending the id as you get it from the route ( so it's either an id or a shortcut )

then, the back-end must be edited like that: 

const filters = Application.find(accessibleBy(ability).Application)
        .where({ _id: args.id })
        .getFilter();

should become

const filters = Application.find(accessibleBy(ability).Application)
        .where({ or: { id: args.id, shortcut: args.shortcut })
        .getFilter();

there is a lot of errors in what I'm putting there, but that's the overall idea
I may be wrong with this approach

Comment (unai-reliefapp) [reply apps/front-office/src/app/app-routing.module.ts]: @AntoineRelief  The idea is to replace the shortcut for the id in the route right?
How im going to send the id and shortcut to the backend if i have access to only one of them from the route on every case?
Following your approach im thinking that we can move the logic in the resolver to the application query and always do the check there, e.g.

```
const isIdShortcut = await Application.findOne({
          shortcut: args.id,
        }).select('id');
const filters = Application.find(accessibleBy(ability).Application)
        .where({ _id: isIdShortcut ?? args.id })
        .getFilter();
```
This way we can also avoid to have a resolver for each application, front, back and web widget(now that the scope has been extended)

Im just trying to follow up

Comment (AntoineRelief) [reply apps/front-office/src/app/app-routing.module.ts]: @unai-reliefapp 
I can't guarantee that would work, but I think the back-end just need to get anything that would identify the application as id argument, would it be the shortcut, or the id. We don't care about that.

Then, you only need to do a query, with a "or" logic, to query both on the id & the shortcut

Comment (AntoineRelief) [reply apps/front-office/src/app/app-routing.module.ts]: so it's kind of what you're telling, but the isIdShortcut seems an extra step

Comment (AntoineRelief) [reply apps/front-office/src/app/app-routing.module.ts]: in case you have some issues doing the query on both fields at the same time, you can indeed use your code, or consider using the mongoose id method to check if the id is valid or not

Comment (unai-reliefapp) [reply apps/front-office/src/app/app-routing.module.ts]: It's working for front and back office, I'll see if i can change the query with that or logic

**Comment by unai-reliefapp** [comment]
Hi @AntoineRelief , do they want the shortcut to be reflected in side urls, e.g.: https://github.com/ReliefApplications/ems-frontend/blob/beta/libs/shared/src/lib/services/download/download.service.ts#L194

Also im assuming to extend the shortcut use to all related places, e.g.
image
If clicking an application and has a shortcut, use it by default, is that correct?

**Comment by AntoineRelief** [comment]
@un

> Hi @AntoineRelief , do they want the shortcut to be reflected in side urls, e.g.: https://github.com/ReliefApplications/ems-frontend/blob/beta/libs/shared/src/lib/services/download/download.service.ts#L194
> 
> Also im assuming to extend the shortcut use to all related places, e.g. image If clicking an application and has a shortcut, use it by default, is that correct?

@unai-reliefapp 
no for the first point, as I believe this is a back-end request so you should not change that

but yes for the front-end navigation, you can use by default the shortcut if we have it

**Comment by unai-reliefapp** [comment]
> @un
> 
> > Hi @AntoineRelief , do they want the shortcut to be reflected in side urls, e.g.: https://github.com/ReliefApplications/ems-frontend/blob/beta/libs/shared/src/lib/services/download/download.service.ts#L194
> > Also im assuming to extend the shortcut use to all related places, e.g. image If clicking an application and has a shortcut, use it by default, is that correct?
> 
> @unai-reliefapp no for the first point, as I believe this is a back-end request so you should not change that
> 
> but yes for the front-end navigation, you can use by default the shortcut if we have it

@AntoineRelief 
That means that by default use the shortcut at any case?
E.g. if i manually use the id to navigate to some application that contains a shortcut, the build up links for widgets would use the shortcut, because it has preference
image
This is acceptable? (I have to navigate using the id manually if shortcut exists for this to happen)

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

yes, it's fine

**Comment by unai-reliefapp** [comment]
Front office error page for application load error
https://github.com/user-attachments/assets/10d98277-61d3-49e7-b0c4-7115347879e8
Back office error page for application load error
https://github.com/user-attachments/assets/7a3b799a-2997-406a-8723-04f84f9ecdd6

@AntoineRelief

**Comment by AntoineRelief** [comment]
@unai-reliefapp 

looks good to me
perhaps we will need to have a button to go back to a default page, but that's out of scope and I need to determine how that would work with WHO, so okay for me like that 👍

**Comment by unai-reliefapp** [comment]
Share feature

https://github.com/user-attachments/assets/8b1938e2-5a32-4269-b75f-7ba135b7b158

All pushed and ready for review @AntoineRelief

File: apps/front-office/src/app/application/application.component.ts [hunk 2]
```diff
@@ -105,15 +104,23 @@ export class ApplicationComponent
     this.applicationService.application$
       .pipe(takeUntil(this.destroy$))
       .subscribe((application: Application | null) => {
+        this.loading = false;
```
Comment (AntoineRelief) [inline apps/front-office/src/app/application/application.component.ts]: @unai-reliefapp 
I have a really quick questino
why did you move this?

Comment (unai-reliefapp) [reply apps/front-office/src/app/application/application.component.ts]: We dont want the loading spinner neither on successful application load, neither on error right?
Whatever the result, stop loading spinner
